src/pauli_exp_val.rs

```diff
@@ -0,0 +1,193 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2022
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use std::convert::TryInto;
+
+use num_complex::Complex64;
+use numpy::PyReadonlyArray1;
+use pyo3::prelude::*;
+use pyo3::wrap_pyfunction;
+use rayon::prelude::*;
+
+use crate::eval_parallel_env;
+
+const LANES: usize = 8;
+
+#[inline]
+fn fast_sum(values: &[f64]) -> f64 {
+    let chunks = values.chunks_exact(LANES);
+    let remainder = chunks.remainder();
+
+    let sum = chunks.fold([0.; LANES], |mut acc, chunk| {
+        let chunk: [f64; LANES] = chunk.try_into().unwrap();
+        for i in 0..LANES {
+            acc[i] += chunk[i];
+        }
+        acc
+    });
+    let remainder: f64 = remainder.iter().copied().sum();
+
+    let mut reduced = 0.;
+    for val in sum.iter().take(LANES) {
```

Comment:
It never would be, my assumption is that this is needed to provide a compiler hint that the iteration is over a fixed width of size `LANES` so it knows that it can use SIMD here too. But I'd have to do A/B testing of the output LLVM IR and assembly with and without `take()` to confirm.

Reply:
Gave it a try just now using Godbolt on 1.59.0. Seems like there's no difference without the `.iter().take(LANES)` (other than which of a few equivalent registers are selected throughout the function).

https://godbolt.org/z/x5bh5PT7j

Reply:
Oh cool, (I always forget about godbolt I always just compile and disassemble by hand locally) then lets just drop the `take()` here if it's not making any difference:

```suggestion
    for val in sum {
```